### ISSUE TITLE
Implement some of the python training/optimization functions

### ISSUE THREAD

**Issue by silky** [tensorflow/haskell#39]
there are many nice functions from `<tf-source>/python/training` that it would be great to have (I'm particularly interested in the adam optimizer, but eventually we'd transfer them all across)

is there some better plan than just porting them?

it might be useful to plan out an interface a bit, because most of them are stateful, and it would be nice to have a clean interaction with sessions.

thoughts?

**Comment by blackgnezdo** [comment]
By stateful, do you mean that they have a TF variable on the side (like accumulators for adagrad)? I implemented a simple case by using [applyAdagrad](https://tensorflow.github.io/haskell/haddock/tensorflow-core-ops-0.1.0.0/TensorFlow-GenOps-Core.html#v:applyAdagrad) directly. Do you have some kind of abstraction in mind which would automatically place the accumulator on the same device and such?

What kind of issues to you anticipate with regard to session interaction?

**Comment by silky** [comment]
well, `adam` has a few updates it performs in the `_finish` method of the python impl (it is, infact, the only one in the core library that does). i was thinking it might be nice to just create a nice abstraction around it all here.

i wasn't really anticipating any issues, i'm just not entirely sure how i'd structure it all. but i'll have a go and see what people think :)

**Comment by blackgnezdo** [comment]
Just in case you need some inspiration, here's a fragment I had around adagrad:
```
    let apply p g =
            let var = paramVar p in
            TF.colocateWith var $ TF.render $
            TF.applyAdagrad var (paramAccum p) learningRate g
    adagrads <- zipWithM apply vars inflatedGrads
    TF.group adagrads
```

**Comment by silky** [comment]
alright.

i've had a rough few hours here, and i'm not sure if i'm up against a bug or just my own incompetence

i thought i could obtain myself an `accum :: Tensor Ref a` by doing something like 

`initializedVariable $ (zerosLike param) + 0.1` where is one of the variables i'm taking gradients of.

but this gives me a runtime error in tensorflow

```
TF_FAILED_PRECONDITION "Attempting to use uninitialized value Variable_13

[[Node: ZerosLike_155 = ZerosLike[T=DT_FLOAT, _device=\"/job:localhost/replica:0/task:0/cpu:0\"](Variable_13)]]
```

i presume it's because param itself is not initialized to a value? and for some reason, zerosLike needs that?

so then I try something like

```
constantLike c t = do
   s <- TF.shape t
   b <- TF.variable s
   w <- TF.assign b (TF.scalar c)
   TF.render w
```

of course, that doesn't work, because `TF.shape` returns a tensor of shapes, not a `TF.Shape` that i can potentially use. so one option would be to *evaluate* that expression in a session, and then build up a new shape, but i'm sure that's a terrible idea...

i believe that in the python there's two ways of obtaining shapes; one that works as an operator, and another one that just lets us obtain the (here) `TF.Shape` that was used to build that particular tensor.

on the other hand, maybe i should be able to do all this in a much much simpler way that i'm completely missing at this point?

thanks.

**Comment by blackgnezdo** [comment]
@silky the issue seems similar to what is solved with [initialized_value](http://stackoverflow.com/questions/34876494/tensorflow-var-eval-vs-var-initialized-value-eval). We don't have a similar mechanism (yet). A workaround would be to create the accumulator explicitly. This unfortunately breaks encapsulation of accumulators in the training op.

While it's easy enough to make the user responsible for initializing things in the correct order, it would be better to prevent the problem at the type level. @judah has some ideas. I'll play with it too.

**Comment by blackgnezdo** [comment]
Repro is trivial:
```
diff --git a/tensorflow-ops/tests/BuildTest.hs b/tensorflow-ops/tests/BuildTest.hs
index 2689dd6..6c76314 100644
--- a/tensorflow-ops/tests/BuildTest.hs
+++ b/tensorflow-ops/tests/BuildTest.hs
@@ -64,6 +64,7 @@ import Test.Framework.Providers.HUnit (testCase)
 import Test.HUnit ((@=?))
 import Google.Test (googleTest)
 import qualified Data.Vector as V
+import qualified TensorFlow.GenOps.Core as CoreOps
 
 -- | Test named behavior.
 testNamed :: Test
@@ -109,7 +110,9 @@ testInitializedVariable =
 testInitializedVariableShape :: Test
 testInitializedVariableShape =
     testCase "testInitializedVariableShape" $ runSession $ do
-        vector <- build $ initializedVariable (constant [1] [42 :: Float])
+        vector <- build $ do
+                  a <- initializedVariable (constant [1] [42 :: Float])
+                  initializedVariable (CoreOps.zerosLike a)
         result <- run vector
         liftIO $ [42] @=? (result :: V.Vector Float)
 
```

**Comment by blackgnezdo** [comment]
Here's what it looks like with [all guts exposed](https://github.com/blackgnezdo/tensorflow-haskell/commit/3f50193954f50dbbb3f773d8f719c887a66825a7). Now the challenge is to make things usable.

**Comment by fkm3** [comment]
I think we can put the initialized value into the new [`Variable`](https://github.com/tensorflow/haskell/blob/master/tensorflow-ops/src/TensorFlow/Variable.hs#L38) type then have a `initializedValue :: Variable a ->  Maybe (Tensor Value a)` function. I'm going to send a CL to add `Variable` support to `gradients` and then try that.

**Comment by silky** [comment]
i'm happy to consider this closed, thanks @fkm3 :)
